text/0000-vec-retain_mut.md

```diff
@@ -0,0 +1,66 @@
+- Feature Name: retain_mut
+- Start Date: 2015-11-6
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a `retain_mut` method to `Vec` and `VecDeque` which allows elements to be
+mutated before deciding whether to retain them.
+
+# Motivation
+[motivation]: #motivation
+
+A common pattern with vectors is to look at each element, modify it and
+remove it from the vector if a certain condition is met. The `retain` method
+allows elements to be removed from a vector in-place but does not allow existing
+elements to be modified.
+
+The lack of a `retain_mut` method means that users have to do 2 passes, one with
+`iter_mut` and one with `retain`. The resulting code is harder to read and
+slower. Another, more efficient, way of doing this would be to work with vector
+indices directly and swap elements into place while mutating them, similar to
+what `retain` already does.
+
+# Detailed design
+[design]: #detailed-design
+
+The only difference between `retain_mut` and `retain` is that `retain_mut` takes
+a `FnMut(&mut T)` parameter instead of a `FnMut(&T)`. This allows elements to
+be mutated before deciding whether to retain them.
+
+The new `retain_mut` method will be added to both `Vec` and `VecDeque`.
+
+Here is an example which will decrement each element of a vector and filter out
+elements that have reached a value of zero.
+
+    let mut vec = vec![7, 1, 3, 10];
+    vec.retain_mut(|&mut x| {
+        x -= 1;
+        x != 0
+    });
```

Comment:
This example won't work as described. the `&mut x` pattern will copy `x` out of the vector and then mutate a stack-local `x` (which might not even compile because I don't think this `x` binding is mutable). The example should be written thusly:

```rust
vec.retain_mut(|x| {
    *x -= 1;
    *x != 0
]);
```